Add documentation about `SeccompDefault` feature (#27957)
kubernetes/website: content/en/docs/tutorials/clusters/seccomp.md

## {{% heading "prerequisites" %}}

{{< version-check >}}

In order to complete all steps in this tutorial, you must install
[kind](https://kind.sigs.k8s.io/docs/user/quick-start/) and
[kubectl](/docs/tasks/tools/). This tutorial will show examples

> with both alpha (pre-v1.22) and generally available seccomp functionality, so

Reviewer: This phrase doesn't make sense. The alpha functionality demonstrated is _not_ pre-v1.22; it actually requires v1.22 to be able to follow every step.

```suggestion
both alpha (new in v1.22) and generally available seccomp functionality. You should
```

make sure that your cluster is [configured
correctly](https://kind.sigs.k8s.io/docs/user/quick-start/#setting-kubernetes-version)
for the version you are using.

<!-- steps -->

## Enable the `RuntimeDefault` seccomp profile for all workloads

> {{< feature-state state="alpha" >}}

Reviewer: ```suggestion
{{< feature-state state="alpha" for_k8s_version="v1.22" >}}
```

A new kubelet alpha [feature gate](/docs/reference/command-line-tools-reference/feature-gates)
`SeccompDefault` has been introduced with the release of Kubernetes v1.22.0.

> If enabled, the kubelet will always use the `RuntimeDefault` seccomp profile of

Reviewer: style nit:
```suggestion
If enabled, the kubelet always uses the `RuntimeDefault` seccomp profile of
```

the corresponding container runtime instead of running the workload in
`Unconfined` (seccomp disabled) mode. Those profiles aim to provide a strong set
of security defaults while preserving the functionality of the workload. It is
possible that the default profiles differ between container runtimes and their
release versions, for example when comparing those from
[CRI-O](https://github.com/cri-o/cri-o/blob/b99366680904420974463572f6f0b92166437b0e/vendor/github.com/containers/common/pkg/seccomp/default_linux.go#L45)
and
[containerd](https://github.com/containerd/containerd/blob/261c107ffc4ff681bc73988f64e3f60c32233b37/contrib/seccomp/seccomp_default.go#L51).

Some workloads may require a lower set of syscall restrictions. This means that
they can fail during runtime even with the `RuntimeDefault` profile. To
mitigate such a failure, either:

- Run the workload explicitly as `Unconfined`.

> - Disable the `SeccompDefault` feature for the nodes.

Reviewer: Readers will also need to make sure that they schedule that workload to the nodes with SeccompDefault disabled. I'd mention that, because AIUI it's fine to have a mixed cluster and label things appropriately (maybe enforcing which namespaces can use which RuntimeClass).

All I'm suggesting here is a hint that you can run a mixed cluster - no need to go into detail!

Reviewer: (nit)

Author: Okay, I mentioned that the user has to make sure that the feature is disabled for the nodes where the workload is scheduled.

- Create a custom seccomp profile for the workload.

> It is recommended to enable the feature on a per-node basis and test the
> execution of workloads in conjunction with the `RuntimeDefault` profile before
> deploying them into production.

Reviewer: ```suggestion
If you were introductiong this feature into production-like cluster, the Kubernetes project
recommends that you enable this feature gate on a subset of your nodes and then
test workload execution before rolling the change out cluster-wide.
```

Since the feature is in alpha state it is disabled per default. To enable it,
pass the flag `--feature-gates=SeccompDefault=true` to the `kubelet` CLI or

> enable it via the [kubelet configuration
> file](/docs/tasks/administer-cluster/kubelet-config-file). To enable the feature

Reviewer: nit:
```suggestion
enable it via the [kubelet configuration
file](/docs/tasks/administer-cluster/kubelet-config-file/). To enable the feature
```

> gate in kind, ensure that it ships the minimum required Kubernetes version as

Reviewer: Either make `kind` a hyperlink or put it in backticks, so it's not confused with the English word “kind”.

Author: Yep, linked it to the default website. :+1:

well as [enabling it in the kind configuration](https://kind.sigs.k8s.io/docs/user/quick-start/#enable-feature-gates-in-your-cluster):

```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  SeccompDefault: true

> ```

Reviewer: If readers can't do this, what should they do (eg: read an older version of this page? Carry on anyway?)

Author: Hm, it's not required to enable the feature for the other parts of the document. Should we outline that to the reader more explicit?

## {{% heading "prerequisites" %}}

{{< version-check >}}

In order to complete all steps in this tutorial, you must install
[kind](https://kind.sigs.k8s.io/docs/user/quick-start/) and
[kubectl](/docs/tasks/tools/). This tutorial will show examples

> both alpha (new in v1.22) and generally available seccomp functionality. You should

Reviewer 2: nit: This tutorial shows examples of alpha (...) and generally available Seccomp functionality. You should ...

Reviewer 2: nit: Check on capitalization of Seccomp (or seccomp).

Author: Fixed and switched to `seccomp`.

make sure that your cluster is [configured
correctly](https://kind.sigs.k8s.io/docs/user/quick-start/#setting-kubernetes-version)
for the version you are using.

<!-- steps -->

> ## Enable the `RuntimeDefault` seccomp profile for all workloads

Reviewer 3: ```suggestion
## Enable the use of `RuntimeDefault` as the default seccomp profile for all workloads
```

{{< feature-state state="alpha" for_k8s_version="v1.22" >}}

> A new optional kubelet [feature

Reviewer 2: Could clean up line 55,
``SeccompDefault` is an optional kubelet [feature gate](...).

gate](/docs/reference/command-line-tools-reference/feature-gates)
`SeccompDefault` has been introduced to Kubernetes.

> If enabled, the kubelet always uses the `RuntimeDefault` seccomp profile of
> the corresponding container runtime instead of running the workload in
> `Unconfined` (seccomp disabled) mode. Those profiles aim to provide a strong set

Reviewer 3: Maybe?
```suggestion
If enabled, the kubelet will use the `RuntimeDefault` seccomp profile by default, which is 
defined by the container runtime, instead  of using the `Unconfined` (seccomp disabled) mode. 
Those profiles aim to provide a strong set
```

of security defaults while preserving the functionality of the workload. It is
possible that the default profiles differ between container runtimes and their
release versions, for example when comparing those from
[CRI-O](https://github.com/cri-o/cri-o/blob/b99366680904420974463572f6f0b92166437b0e/vendor/github.com/containers/common/pkg/seccomp/default_linux.go#L45)
and
[containerd](https://github.com/containerd/containerd/blob/261c107ffc4ff681bc73988f64e3f60c32233b37/contrib/seccomp/seccomp_default.go#L51).

Some workloads may require a lower set of syscall restrictions. This means that
they can fail during runtime even with the `RuntimeDefault` profile. To
mitigate such a failure, either:

- Run the workload explicitly as `Unconfined`.
- Disable the `SeccompDefault` feature for the nodes. Also making sure that
  workloads get scheduled on nodes where the feature is disabled.
- Create a custom seccomp profile for the workload.

> If you were introductiong this feature into production-like cluster, the Kubernetes project

Reviewer 3: ```suggestion
If you were introducing this feature into production-like cluster, the Kubernetes project
```

> If you were introductiong this feature into production-like cluster, the Kubernetes project

Reviewer 2: Possibly reword:
For production systems, the Kubernetes project recommends that you enable the `SeccompDefault` feature gate on ...

recommends that you enable this feature gate on a subset of your nodes and then
test workload execution before rolling the change out cluster-wide.

Since the feature is in alpha state it is disabled per default. To enable it,
pass the flag `--feature-gates=SeccompDefault=true` to the `kubelet` CLI or
enable it via the [kubelet configuration
file](/docs/tasks/administer-cluster/kubelet-config-file/). To enable the feature

> gate in [kind](https://kind.sigs.k8s.io), ensure that it ships the minimum required Kubernetes version as

Reviewer 2: nit: Possibly reword (ensure that it ships ...) to something like:
To enable the feature gate in [kind](...), ensure that `kind` integrates with [or provides] Kubernetes version {{< param "version" >}} or higher.

## {{% heading "prerequisites" %}}

{{< version-check >}}

In order to complete all steps in this tutorial, you must install
[kind](https://kind.sigs.k8s.io/docs/user/quick-start/) and
[kubectl](/docs/tasks/tools/). This tutorial will show examples
both alpha (new in v1.22) and generally available seccomp functionality. You should
make sure that your cluster is [configured
correctly](https://kind.sigs.k8s.io/docs/user/quick-start/#setting-kubernetes-version)
for the version you are using.

<!-- steps -->

## Enable the use of `RuntimeDefault` as the default seccomp profile for all workloads

{{< feature-state state="alpha" for_k8s_version="v1.22" >}}

`SeccompDefault` is an optional kubelet
[feature gate](/docs/reference/command-line-tools-reference/feature-gates) as
well as corresponding `--seccomp-default`
[command line flag](/docs/reference/command-line-tools-reference/kubelet).
Both have to be enabled simultaneously to use the feature.

If enabled, the kubelet will use the `RuntimeDefault` seccomp profile by default, which is 
defined by the container runtime, instead  of using the `Unconfined` (seccomp disabled) mode. 

> Those profiles aim to provide a strong set

Author: Changed to "The default"

of security defaults while preserving the functionality of the workload. It is
possible that the default profiles differ between container runtimes and their
release versions, for example when comparing those from
[CRI-O](https://github.com/cri-o/cri-o/blob/b99366680904420974463572f6f0b92166437b0e/vendor/github.com/containers/common/pkg/seccomp/default_linux.go#L45)
and

> [containerd](https://github.com/containerd/containerd/blob/261c107ffc4ff681bc73988f64e3f60c32233b37/contrib/seccomp/seccomp_default.go#L51).

Author: Alright, removed the links.

> Some workloads may require a lower set of syscall restrictions. This means that

Author: Rephrased to: "Some workloads may require a lower amount of syscall restrictions than others"

<!-- steps -->

## Enable the `RuntimeDefault` seccomp profile for all workloads

A new kubelet alpha [feature gate](/docs/reference/command-line-tools-reference/feature-gates)

> `SeccompDefault` has been introduced with the release of Kubernetes v1.22.0.

Reviewer: This should use a [feature state shortcode](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#feature-state).

Author: Added a feature short code for this section

Reviewer: With the feature state line in place, you can omit this one. It should come out because phrases like this are relevant at the point of release but don't age well, whereas a feature-state marker is valid until the facts change.

(This is different from blog articles which absolutely can and do use wording similar to the line I'm commenting on).
